src/lib.ts

```diff
@@ -0,0 +1,343 @@
+import { S3Client, S3ClientConfig, GetObjectCommand, GetObjectCommandOutput } from '@aws-sdk/client-s3';
+import imageminPngquant from 'imagemin-pngquant';
+import sharp from 'sharp';
+import smartcrop from 'smartcrop-sharp';
+
+export interface Args {
+	// Required args.
+	key: string;
+
+	// Optional args.
+	background?: string;
+	crop?: string | string[];
+	crop_strategy?: string;
+	fit?: string;
+	gravity?: string;
+	h?: string;
+	lb?: string;
+	resize?: string | number[];
+	quality?: string | number;
+	w?: string;
+	webp?: string | boolean;
+	zoom?: string;
+	'X-Amz-Algorithm'?: string;
+	'X-Amz-Content-Sha256'?: string;
+	'X-Amz-Credential'?: string;
+	'X-Amz-SignedHeaders'?: string;
+	'X-Amz-Expires'?: string;
+	'X-Amz-Signature'?: string;
+	'X-Amz-Date'?: string;
+	'X-Amz-Security-Token'?: string;
+}
+
+export type Config = S3ClientConfig & { bucket: string };
+
+/**
+ * Get the dimensions from a string or array of strings.
+ */
+function getDimArray( dims: string | number[], zoom: number = 1 ): ( number | null )[] {
+	let dimArr = typeof dims === 'string' ? dims.split( ',' ) : dims;
+	return dimArr.map( v => Math.round( Number( v ) * zoom ) || null );
+}
+
+/**
+ * Clamp a value between a min and max.
+ */
+function clamp( val: number | string, min: number, max: number ): number {
+	return Math.min( Math.max( Number( val ), min ), max );
+}
+
+/**
+ * Get a file from S3/
+ */
+export async function getS3File( config: Config, key: string, args: Args ): Promise<GetObjectCommandOutput> {
+	const s3 = new S3Client( {
+		...config,
+		signer: {
+			/**
+			 *
+			 * @param request
+			 */
+			sign: async request => {
+				if ( ! args['X-Amz-Algorithm'] ) {
+					return request;
+				}
+				const presignedParamNames = [
+					'X-Amz-Algorithm',
+					'X-Amz-Content-Sha256',
+					'X-Amz-Credential',
+					'X-Amz-SignedHeaders',
+					'X-Amz-Expires',
+					'X-Amz-Signature',
+					'X-Amz-Date',
+					'X-Amz-Security-Token',
+				] as const;
+				const presignedParams: { [K in ( typeof presignedParamNames )[number]]?: string } = {}; // eslint-disable-line no-unused-vars
+				const signedHeaders = args['X-Amz-SignedHeaders']?.split( ';' ) || [];
+
+				for ( const paramName of presignedParamNames ) {
+					if ( args[paramName] ) {
+						presignedParams[paramName] = args[paramName];
+					}
+				}
+
+				const headers: typeof request.headers = {};
+				for ( const header in request.headers ) {
+					if ( signedHeaders.includes( header.toLowerCase() ) ) {
```

Comment:
Does `signedHeaders` need a `.map( t => t.toLowerCase() )` as well? (And maybe a trim?)

Reply:
Added a lowercase and trim here